File: backend/app/schemas/statistika.py
from datetime import date

from pydantic import BaseModel


class MahsulotJamlanmasi(BaseModel):
    mahsulot_kodi: str
    mahsulot_nomi: str
    soni: int
    jami_kg: float


class DavrJamlanmasi(BaseModel):
    davr: str
    boshlanish_sanasi: date
    tugash_sanasi: date
    mahsulotlar: list[MahsulotJamlanmasi]
    jami_soni: int
    jami_kg: float


class SmenaJamlanmasi(BaseModel):
    smena: str
    soni: int
    jami_kg: float


class OperatorJamlanmasi(BaseModel):
    operator_id: int
    ism: str
    login: str
    smena: str | None
    soni: int
    jami_kg: float


class RekordSmena(BaseModel):
    smena: str
    jami_kg: float


class RekordOperator(BaseModel):
    operator_id: int
    ism: str
    login: str
    soni: int


class RekordKun(BaseModel):
    sana: date
    jami_kg: float


class RekordlarJavob(BaseModel):
    """Statistika ekranidagi "Rekord" paneli uchun. eng_yaxshi_smena va
    eng_yaxshi_operator tanlangan `davr` bo'yicha; eng_yuqori_kunlik_yigim
    esa DOIM barcha vaqt bo'yicha (davrga bog'liq emas)."""

    davr: str
    boshlanish_sanasi: date
    tugash_sanasi: date
    eng_yaxshi_smena: RekordSmena | None
    eng_yaxshi_operator: RekordOperator | None
    eng_yuqori_kunlik_yigim: RekordKun | None

File: frontend/lib/models/statistika.dart
class DavrJamlanmasi {
  final String davr;
  final String boshlanishSanasi;
  final String tugashSanasi;
  final List<MahsulotJamlanmasi> mahsulotlar;
  final int jamiSoni;
  final double jamiKg;

  DavrJamlanmasi({
    required this.davr,
    required this.boshlanishSanasi,
    required this.tugashSanasi,
    required this.mahsulotlar,
    required this.jamiSoni,
    required this.jamiKg,
  });

  factory DavrJamlanmasi.fromJson(Map<String, dynamic> j) => DavrJamlanmasi(
        davr: j['davr'],
        boshlanishSanasi: j['boshlanish_sanasi'],
        tugashSanasi: j['tugash_sanasi'],
        mahsulotlar: (j['mahsulotlar'] as List).map((e) => MahsulotJamlanmasi.fromJson(e)).toList(),
        jamiSoni: j['jami_soni'],
        jamiKg: (j['jami_kg'] as num).toDouble(),
      );
}

class MahsulotJamlanmasi {
  final String mahsulotKodi;
  final String mahsulotNomi;
  final int soni;
  final double jamiKg;

  MahsulotJamlanmasi({required this.mahsulotKodi, required this.mahsulotNomi, required this.soni, required this.jamiKg});

  factory MahsulotJamlanmasi.fromJson(Map<String, dynamic> j) => MahsulotJamlanmasi(
        mahsulotKodi: j['mahsulot_kodi'],
        mahsulotNomi: j['mahsulot_nomi'],
        soni: j['soni'],
        jamiKg: (j['jami_kg'] as num).toDouble(),
      );
}
